Simulator/Utils/DeviceVector.h

```diff
@@ -0,0 +1,439 @@
+/**
+ * @file DeviceVector.h
+ *
+ * @ingroup Simulator/Utils
+ *
+ * @brief A Template class that provides a std::vector-like interface for both host and device (GPU) memory
+ * @tparam T Type of elements (must be BGFLOAT, int, or bool)
+ *
+ * This class wraps a std::vector and provides additional operations for device memory
+ * allocation, deallocation, and host-device data transfer. Currently, it only supports
+ * primitive types, and since the codebase only uses BGFLOAT, int, and bool, these are
+ * the only types explicitly allowed. Support for other primitive types (like char,
+ * double, etc.) can be added if needed in the future. Note that
+ * this class is not RAII compliant for device memory - it provides methods for manual
+ * device memory management but leaves the responsibility of proper allocation,
+ * synchronization, and deallocation to the user.
+ *
+ * The class provides convenient methods for common host-device operations but performs
+ * only trivial safety checks. Users must ensure device memory operations are safe and
+ * that host-device memory is properly synchronized using the provided methods. For example,
+ * operations like copyToDevice() or copyToHost() will not implicitly allocate device
+ * memory if it hasn't been allocated - they will throw an error instead. Similar behavior
+ * applies to other device operations.
+ *
+ * Host-side operations are implemented by overloading std::vector operations and operate
+ * independently of device memory - they do not automatically sync changes to the device.
+ * For detailed behavior of these operations, refer to std::vector documentation. Future
+ * needs for additional vector operations can be met by overloading the specific
+ * methods or operators as needed.
+ 
+ * Currently, DeviceVector only supports primitive types, and since the codebase only uses
+ * BGFLOAT, int, and bool, these are the only types explicitly allowed. Support for other
+ * primitive types (like char, double, etc.) can be added if needed in the future.
+ *
+ * Note: The current implementation prioritizes explicit control over convenience.
+ * Future extensions could add RAII compliance, additional safety checks, and automatic
+ * host-device synchronization if such behavior becomes desirable.
+ */
+
+#pragma once
+#include "BGTypes.h"   // For BGFLOAT definition
+#include <stdexcept>
+#include <vector>
+
+#if defined(__CUDACC__)
+   #include "Book.h"
+#endif
+
+#include <variant>
+
+/// Define supported types for DeviceVector using variant
+using DeviceVectorTypes = std::variant<BGFLOAT, int, bool>;
+
+/// Type trait to check if a type is one of the alternatives in a variant
+template <typename T, typename Variant> struct is_device_vector_type : std::false_type {};
+
+template <typename T, typename... Types>
+struct is_device_vector_type<T, std::variant<Types...>> :
+   std::bool_constant<(std::is_same_v<T, Types> || ...)> {};
```

Comment:
I hate to do this, but could you include more explanation of this in the comment? For example, the obvious first question is why doesn't this trigger a "multiply defined" compiler error for `struct is_device_vector_type`?